Only change to short IDs for delete

abadger opened the pull request:
##### ISSUE TYPE
- Bugfix Pull Request
##### COMPONENT NAME

apt_key.py
##### ANSIBLE VERSION

<!--- Paste verbatim output from “ansible --version” between quotes below -->

```
devel stable-2.2
```
##### SUMMARY

If the user specifies long IDs, use them for all commands except for
deleting a key.  Need to use short IDs there because of an upstream
apt_key bug.  Fixed in apt_key 1.10 (fix is present in Ubuntu 16.04 but
not Ubuntu 14.0 or some Debians).

Fixes #5237

Also:
- Some style cleanups
- Return a nice user error message if the key was not found on the
  keyserver
- Make file and keyring parameters type='path' so envars and tilde are
  expanded

abadger committed:
Only change to short IDs for delete

If the user specifies long IDs, use them for all commands except for
deleting a key.  Need to use short IDs there because of an upstream
apt_key bug.  Fixed in apt_key 1.10 (fix is present in Ubuntu 16.04 but
not Ubuntu 14.0 or some Debians).

Fixes #5237

Also:

* Some style cleanups
* Return a nice user error message if the key was not found on the
  keyserver
* Make file and keyring parameters type='path' so envars and tilde are
  expanded

gregdek commented:
Thanks @abadger. To the current maintainers, @jvantuyl please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with text 'shipit', 'needs_revision' or 'close_me' as appropriate.

[This message brought to you by your friendly Ansibull-bot.]

abadger committed:
Check that apt-key really erased the key

When erasing a key, apt-key does not understand how to process subkeys.
This update explicitly checks that the key_id is no longer present and
throws an error if it is.  It also hints at subkeys being a possible
problem in the error message and the documentation.

Fixes #5119

abadger committed:
Fix apt_key check mode with long ids

* More cleanups
* apt-key can be given a key id longer than 16 chars to more accurately
  define what key to download.  However, we can use a maximum of 16
  chars to verify whether a key is installed or not.  So we need to use
  different lengths for the id depending on what we're doing with it.

Fixes #2622

abadger commented:
Merged to devel and stbale-2.2.  This should be in the 2.2.0 release.

ypid commented on packaging/os/apt_key.py:
```diff
@@ -237,64 +284,70 @@ def main():
-            if key_id.startswith('0x'):
-                key_id = key_id[2:]
-            key_id = key_id.upper()[-8:]
+            key_id, fingerprint, short_key_id = parse_key_id(key_id)
```
@abadger You mixed up the return order. It should be: `short_key_id, fingerprint, key_id = parse_key_id(key_id)`

ypid commented:
@abadger Thanks but even after the return fix it still showed ok for this task when the key is already present (the short key id only obvious):

``` YAML
- apt_key:
    # id: 'DDA2C105C4B73A6649AD2BBD47AE7F72479BC94B'
    id: '00000000000000000000000047AE7F72479BC94B'
  keyserver: 'hkp://pool.sks-keyservers.net'
```

Can someone else also review this? cc @jvantuyl

abadger commented:
Fixed the problem ypid identified here: https://github.com/ansible/ansible-modules-core/pull/5357

If we decide to backport this to stable-2.1, we'll need to cherry-pick that commit as well.
